modules/Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/SubcraftValidator.java

```diff
@@ -0,0 +1,39 @@
+package net.countercraft.movecraft.processing.tasks.detection;
+
+import net.countercraft.movecraft.Movecraft;
+import net.countercraft.movecraft.MovecraftLocation;
+import net.countercraft.movecraft.craft.Craft;
+import net.countercraft.movecraft.craft.CraftManager;
+import net.countercraft.movecraft.craft.CraftType;
+import net.countercraft.movecraft.localisation.I18nSupport;
+import net.countercraft.movecraft.processing.MovecraftWorld;
+import net.countercraft.movecraft.processing.TaskPredicate;
+import net.countercraft.movecraft.processing.WorldManager;
+import org.bukkit.Bukkit;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Supplier;
+
+public class SubcraftValidator implements TaskPredicate<MovecraftLocation> {
+    @Override
+    public Result validate(@NotNull MovecraftLocation movecraftLocation, @NotNull CraftType type, @NotNull MovecraftWorld world, @Nullable CommandSender player) {
+
+        if (!type.getMustBeSubcraft()) {
+            return Result.succeed();
+        }
+
+        //This has to be executed on the main thread because otherwise it could cause deadlock
+        return WorldManager.INSTANCE.executeMain(new Supplier<Result>() {
+            @Override
+            public Result get() {
+                Craft otherCraft = CraftManager.getInstance().fastNearestCraftToLoc(movecraftLocation.toBukkit(Bukkit.getWorld(world.getWorldUUID())));
+                if (otherCraft == null) {
+                    return Result.failWithMessage(I18nSupport.getInternationalisedString("Detection - Must Be Subcraft"));
+                }
+                return (otherCraft.getHitBox().contains(movecraftLocation) ? Result.succeed() : Result.failWithMessage(I18nSupport.getInternationalisedString("Detection - Must Be Subcraft")));
+            }
+        });
```

Comment:
Supplier is a [`FunctionalInterface`](https://docs.oracle.com/javase/8/docs/api/java/lang/FunctionalInterface.html), and thus can be supplanted by a lambda expression, however when using `executeMain` we should consider why it is done. Anything that calls the method becomes a critical path in the threading model for the plugin, so we should avoid it when possible. In this case, its because `fastNearestCraftToLoc` is being invoked, which warns the user about the thread safety of its execution context - however in reality it is perfectly safe. The issue arises because of an internal call to `Craft::getWorld`, which could be replaced by `Craft::getMovecraftWorld`, thus allowing us to run this validator fully off thread. If you'd like you can try to fix this or you can wait on me to do so.